### ISSUE TITLE
Talisman pre-commit hook

### ISSUE THREAD

**Issue by Puneet1726** [thoughtworks/talisman#165]
I am running talisman as pre-commit hook. I have added a text file with password in it but during commit it did not raise any alert.

Please let me know whether it checks password or not. I ram talisman on Linux instance
PFA is the log.

[XXXXXXXXXXXXX gittest]# git add Test.txt
[XXXXXXXXXXXXX gittest]# git commit Test.txt -m "test"
[master faf918b] test
 1 file changed, 1 insertion(+), 1 deletion(-)
[XXXXXXXXXXXXX gittest]# ls -lrt
total 8
drwxr-xr-x 3 root root  18 Oct 29 10:40 talisman_reports
drwxr-xr-x 2 root root   6 Oct 29 10:42 talisman_html_report
-rw-r--r-- 1 root root  15 Oct 29 10:44 Test.txt
[XXXXXXXXXXXXX gittest]# cat Test.txt
passwords=1234
[XXXXXXXXXXXXX gittest]#

**Comment by jpninanjohn** [comment]
I can confirm this. Think this is because the patterns we check do not include passwords (plural form).

**Comment by mrngm** [comment]
In `filename_detector.go`, there is a pattern that literally says `password`, therefore any filename that contains the word `password` will match.

In `pattern_detector.go`, there are extra rules for what should surround the word `password` (as of commit dac0d04):
```
pattern_detector.go:		regexp.MustCompile("(?i)(['\"_]?password['\"]? *[:=][^,;\n]{8,})"),
pattern_detector.go:		regexp.MustCompile("(<[^(><.)]?password[^(><.)]*?>[^(><.)]+</[^(><.)]?password[^(><.)]*?>)"),
```
such as `'`, `"` optionally as prefix or suffix, but if they're not present, only `password ` (note the space` will match for the first pattern, or `password>` for the second one.

**Comment by Puneet1726** [comment]
@mrngm @jpninanjohn : So you are saying it will check only the filename with password and not inside the file?

**Comment by mrngm** [comment]
There is another thing hidden in the pattern, it will only match when there are _at least_ 8 characters after `=` and they must not be `,`, `;` or a newline.

As to why a default configuration would not scan files, I have no idea.

**Comment by Puneet1726** [comment]
@jpninanjohn @mrngm : It should check inside the file. Password will be inside the file

**Comment by jpninanjohn** [comment]
> There is another thing hidden in the pattern, it will only match when there are _at least_ 8 characters after `=` and they must not be `,`, `;` or a newline.
> 
> As to why a default configuration would not scan files, I have no idea.

You are right. It works for text larger than 8 letters. Why is the length of the password considered?

**Comment by mrngm** [comment]
@Puneet1726 Could you run a new test, but using `TALISMAN_DEBUG="some-non-empty-value" git commit -m "your msg" file.txt` when committing?

That should give some more information when running the hook. If you see no output, check the permissions of your `/root/.talisman/bin/talisman_hook_script`. They should have the executable bit set.

**Comment by mrngm** [comment]
> > There is another thing hidden in the pattern, it will only match when there are _at least_ 8 characters after `=` and they must not be `,`, `;` or a newline.
> > As to why a default configuration would not scan files, I have no idea.
> 
> You are right. It works for text larger than 8 letters. Why is the length of the password considered?

Perhaps @harinee or @jacksingleton can comment on that.

**Comment by harinee** [comment]
I do think this would be a bug if it is not looking into the content of a file.

> > > There is another thing hidden in the pattern, it will only match when there are _at least_ 8 characters after `=` and they must not be `,`, `;` or a newline.
> > > As to why a default configuration would not scan files, I have no idea.
> > 
> > 
> > You are right. It works for text larger than 8 letters. Why is the length of the password considered?
> 
> Perhaps @harinee, @jacksingleton, or the original committer @prachisr can comment on that.

This was because password policies in production systems typically impose a check of minimum 8 characters.
The worry with introducing 'password' as a check in Talisman was because of the numerous places where it would cry out false positives in the code.
So some constraints were put in place to try and detect if it would really be a password entered.
Open to hear feedback on that though. (FYI, we have got some feedback already that even the current check throws out too many false positives. But no one raised any issues on github officially I think)

**Comment by jpninanjohn** [comment]
I think the length of a password does not matter. Atleast for passwords, even if its a false positive, the user can ignore it explicitly.
